From 95866cace6d5e1dd6e810f18e12969473fad2aa2 Mon Sep 17 00:00:00 2001
From: Grygorii Strashko <grygorii_strashko@epam.com>
Date: Wed, 24 Apr 2024 15:10:10 +0300
Subject: [PATCH] drivers: mmc: host: sdhci-brcmstb: fix no pinctrl case

In no pinctrl case, when pinctrl are set by u-boot for example, driver
check for pictrl presence and sets no_pinctrl variable if not present, but
then still trying to acceess pinctrl device which causes crash.

Fix sdhci-brcmstb to handle no pinctrl case properly.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Upstream-Status: Inappropriate [downstream fork / BSP-layer delta; not submitted upstream]
Origin: commit 95866cace6d5e1dd6e810f18e12969473fad2aa2
---
 drivers/mmc/host/sdhci-brcmstb.c | 21 +++++++++++----------
 1 file changed, 11 insertions(+), 10 deletions(-)

diff --git a/drivers/mmc/host/sdhci-brcmstb.c b/drivers/mmc/host/sdhci-brcmstb.c
index 61cea31cc891..13b522031fb2 100644
--- a/drivers/mmc/host/sdhci-brcmstb.c
+++ b/drivers/mmc/host/sdhci-brcmstb.c
@@ -553,16 +553,17 @@ static int sdhci_brcmstb_probe(struct platform_device *pdev)
 	priv->pinctrl = devm_pinctrl_get(&pdev->dev);
 	if (IS_ERR(priv->pinctrl)) {
 			no_pinctrl = true;
-	}
-	priv->pins_default = pinctrl_lookup_state(priv->pinctrl, "default");
-	if (IS_ERR(priv->pins_default)) {
-			dev_dbg(&pdev->dev, "No pinctrl default state\n");
-			no_pinctrl = true;
-	}
-	priv->pins_sdex = pinctrl_lookup_state(priv->pinctrl, "sd-express");
-	if (IS_ERR(priv->pins_sdex)) {
-			dev_dbg(&pdev->dev, "No pinctrl sd-express state\n");
-			no_pinctrl = true;
+	} else {
+		priv->pins_default = pinctrl_lookup_state(priv->pinctrl, "default");
+		if (IS_ERR(priv->pins_default)) {
+				dev_dbg(&pdev->dev, "No pinctrl default state\n");
+				no_pinctrl = true;
+		}
+		priv->pins_sdex = pinctrl_lookup_state(priv->pinctrl, "sd-express");
+		if (IS_ERR(priv->pins_sdex)) {
+				dev_dbg(&pdev->dev, "No pinctrl sd-express state\n");
+				no_pinctrl = true;
+		}
 	}
 	if (no_pinctrl || !priv->sde_ioaddr || !priv->sde_ioaddr2) {
 		priv->pinctrl = NULL;
-- 
2.34.1
